internal/Yesod/Session/Memcache/Storage.hs

```diff
@@ -0,0 +1,91 @@
+module Yesod.Session.Memcache.Storage
+  ( memcacheStorage
+  , SessionPersistence (..)
+  ) where
+
+import Internal.Prelude
+
+import Control.Monad.Reader (MonadReader (ask))
+import Database.Memcache.Client qualified as Memcache
+import Database.Memcache.Types qualified as Memcache
+import Session.Key
+import Yesod.Session.SessionType
+import Yesod.Session.Storage.Exceptions
+import Yesod.Session.Storage.Operation
+
+-- | Mapping between 'Session' and Memcache representation.
+data SessionPersistence = SessionPersistence
+  { databaseKey :: SessionKey -> Memcache.Key
+  , toDatabase :: Session -> Memcache.Value
+  , fromDatabase :: Memcache.Value -> Session
+  , runDB :: forall a b m. ReaderT b IO a -> m a
+  }
+
+memcacheStorage
+  :: forall m result
+   . (MonadReader Memcache.Client m, MonadIO m)
+  => SessionPersistence
+  -> StorageOperation result
+  -> m result
+memcacheStorage sp@SessionPersistence {} = \case
+  GetSession sessionKey ->
+    let get client = Memcache.get client (sp.databaseKey sessionKey)
+    in  (sp.fromDatabase . fstOf3) <$$> (ask >>= liftIO . get)
+  DeleteSession sessionKey ->
+    ask
+      >>= \client ->
+        liftIO
+          $ Memcache.delete client (sp.databaseKey sessionKey) bypassCAS
+          >>= \success -> unless success $ throwWithCallStack FailedToDeleteSession
+  InsertSession session ->
+    let
+      key = sp.databaseKey session.key
+      value = sp.toDatabase session
+    in
+      ask
+        >>= \client ->
+          liftIO
+            $ Memcache.add client key value defaultFlags cacheForever
+            >>= \case
+              Nothing -> throwWithCallStack SessionAlreadyExists
+              Just _ -> pure ()
+  ReplaceSession session ->
+    let key = sp.databaseKey session.key
+    in  ask
+          >>= \client ->
+            liftIO
+              $ Memcache.replace
+                client
+                key
+                (sp.toDatabase session)
+                defaultFlags
+                cacheForever
+                bypassCAS
+              >>= \case
+                Nothing -> throwWithCallStack SessionDoesNotExist
+                Just _ -> pure ()
+
+-- | Do not expire the session via Memcache expiration.
+--
+-- Not all storage backends support expiration of keys. Furthermore, we want to
+-- rely on /timing/ fields in the 'Options' data type to determine when a session expires.
+--
+-- N.B. No garbage collection of expired sessions is performed by this library.
+cacheForever :: Memcache.Expiration
+cacheForever = 0
```

Comment:
I like that memcache can clean up automatically, and I like that this approach is robust against changing the server configuration to extend the timeout durations, but - We should set expiration times. Since we're using memcache as the sole source of truth here and not merely as a cache, ops has to monitor the storage and get alerts when we're nearing capacity to make sure we don't overflow it and start dropping sessions.

Might we worth it to make the library configurable - our memcache `SessionPersistence` type could have a field of some type like

```haskell
data MemcacheExpiration =
  -- | Do not set expiration times; memache will only evict when the cache is full
  NoMemcacheExpiration |
  -- | Sessions will be stored in memcache with the same expiration time that we
  -- send to the HTTP client, the lesser of the idle and absolute timeouts.
  UseMemcacheExpiration
```